libsolidity/formal/CHC.cpp

```diff
@@ -541,6 +541,48 @@ void CHC::endVisit(Return const& _return)
 	m_currentBlock = predicate(*returnGhost);
 }
 
+bool CHC::visit(TryStatement const& _tryStatement)
+{
+	FunctionCall const* externalCall = dynamic_cast<FunctionCall const*>(&_tryStatement.externalCall());
+	solAssert(externalCall && externalCall->annotation().tryCall, "");
+	solAssert(m_currentFunction, "");
+
+	auto tryHeaderBlock = createBlock(&_tryStatement, PredicateType::FunctionBlock, "try_header_");
+	auto afterTryBlock = createBlock(&m_currentFunction->body(), PredicateType::FunctionBlock);
+
+	auto const& clauses = _tryStatement.clauses();
+	solAssert(clauses[0].get() == _tryStatement.successClause(), "First clause of TryStatement should be the success clause");
+	auto clauseBlocks = applyMap(clauses, [this](ASTPointer<TryCatchClause> clause) {
+		return createBlock(clause.get(), PredicateType::FunctionBlock, "try_clause_" + std::to_string(clause->id()));
+	});
+
+	connectBlocks(m_currentBlock, predicate(*tryHeaderBlock));
+	setCurrentBlock(*tryHeaderBlock);
+	// Visit everything, except the actual external call.
+	externalCall->expression().accept(*this);
+	ASTNode::listAccept(externalCall->arguments(), *this);
+	// Branch directly to all catch clauses, since in these cases, any effects of the external call are reverted.
+	for (size_t i = 1; i < clauseBlocks.size(); ++i)
+		connectBlocks(m_currentBlock, predicate(*clauseBlocks[i]));
+	// Only now visit the actual call to record its effects and connect to the success clause.
+	endVisit(*externalCall);
+	if (_tryStatement.successClause()->parameters())
+		tryCatchAssignment(
+			_tryStatement.successClause()->parameters()->parameters(), *m_context.expression(*externalCall)
+		);
+	connectBlocks(m_currentBlock, predicate(*clauseBlocks[0]));
```

Comment:
Just as a note: We could probably still simplify this further by just directly using
``clauses[0]->accept(*this); connectBlocks(m_currentBlock, predicate(*afterTryBlock));`` here and starting the loop below at ``1`` instead (respectively in the end generating one clause block less)...

But maybe it's simpler and the predicates are easier to read later, if we keep it like this, so fine with me either way.

Reply:
I see, I agree that it should be possible.

The disadvantage is that we lose symmetry: catch clauses would have a dedicated predicate, but the success clause would not have one. And the success case might be harder to follow in the encoding.

The advantage is that we do less work and there is one less predicate in the encoding, which might be better for the solver.
On the other hand, having the predicate might make extracting and interpreting a counterexample easier.

I am currently slightly in favour of the current version, with the explicit success block.